source/faucet/main.d

```diff
@@ -272,6 +273,81 @@ public class Faucet : IFaucet
     {
         return this.state.utxos.storage;
     }
+
+    // Get UTXOs owned by us
+    private UTXO[Hash] getOwnedUTXOs () nothrow @safe
+    {
+        return this.state.utxos.storage.byKeyValue()
+                   .filter!(
+                       kv => kv.value.output.address == WK.Keys[kv.value.output.address].address)
+                   .map!(kv => tuple(kv.key, kv.value))
+                   .assocArray();
+    }
+
+    /// POST: /send_transaction
+    public override void sendTransaction (string recv, ulong amount)
+    {
+        PublicKey pubkey = PublicKey.fromString(recv);
+        Amount leftover = Amount(amount);
+        auto first_utxo = this.getOwnedUTXOs().byKeyValue().take(1);
+        // A storage to keep track of used UTXOs
+        UTXO[Hash] used_utxos;
+
+        assert(first_utxo.map!(utxo => utxo.value.output.value).front > Amount(0));
+
+        TxBuilder txb = TxBuilder(first_utxo.map!(utxo => utxo.value.output)
+                                            .zip(first_utxo.map!(
+                                                 utxo => utxo.key))
+                                            .front.expand);
+
+        used_utxos[first_utxo.map!(utxo => utxo.key).front]
+            = first_utxo.map!(utxo => utxo.value).front;
+
+        if (leftover <= first_utxo.map!(utxo => utxo.value.output.value).front)
+        {
+            Transaction tx = txb.draw(leftover, [pubkey]).sign();
+            logInfo("Sending %s BOA to %s", amount, recv);
+            this.client.putTransaction(tx);
+        }
+
+        else
+        {
+            txb.draw(first_utxo.map!(utxo => utxo.value.output.value).front, [pubkey]);
+            leftover.sub(first_utxo.map!(utxo => utxo.value.output.value).front);
+
+            while (leftover > Amount(0))
+            {
+                assert(this.getOwnedUTXOs().length);
+
+                auto new_utxo = this.getOwnedUTXOs().byKeyValue()
```

Comment:
I think you're using the wrong approach here. Specifically in the previous code you had:

```
auto first_utxo = this.getOwnedUTXOs().byKeyValue().take(1);
```

Instead what you should do at the beginning is:

```D
auto owned_utxos = this.getOwnedUTXOs().byKeyValue()
    .filter!(pair => pair.key !in used_utxos);
auto first_utxo = owned_utxos.front;
owned_utxos.popFront();
```

And `used_utxos` needs to become a field, not a variable on the stack.

And then at this line 322 you would simply pop from `owned_utxos` instead of calling `getOwnedUTXOs()`.